development/gcbuild/main.go

```diff
@@ -0,0 +1,340 @@
+	"flag"
+	"fmt"
+	"github.com/kyma-project/test-infra/development/gcbuild/tags"
+	"gopkg.in/yaml.v2"
+	"io"
+	"io/ioutil"
+	errutil "k8s.io/apimachinery/pkg/util/errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+type Config struct {
+	Steps         []Step            `yaml:"steps"`
+	Substitutions map[string]string `yaml:"substitutions"`
+	Images        []string          `yaml:"images"`
+}
+
+func getConfig(f string) (*Config, error) {
+	b, err := ioutil.ReadFile(f)
+	if err != nil {
+		return nil, err
+	}
+	var cb Config
+	if err := yaml.Unmarshal(b, &cb); err != nil {
+		return nil, fmt.Errorf("cloudbuild.yaml parse error: %w", err)
+	}
+	return &cb, nil
+}
+
+type Step struct {
+	Name string   `yaml:"name"`
+	Args []string `yaml:"args"`
+}
+
+type options struct {
+	buildDir      string
+	configFile    string
+	variantsFile  string
+	variant       string
+	logDir        string
+	devRegistry   string
+	project       string
+	stagingBucket string
+	logsBucket    string
+	silent        bool
+	isCI          bool
+}
+
+// parseVariable returns a gcloud compatible substitution option.
+// Keys are set to upper-case and prefix "_" is set if not present.
+func parseVariable(key, value string) string {
+	k := strings.TrimSpace(strings.ToUpper(key))
+	if !strings.HasPrefix(k, "_") {
+		k = "_" + k
+	}
+	return k + "=" + strings.TrimSpace(value)
+}
+
+func getSHAFromGit() (string, error) {
+	cmd := exec.Command("git", "rev-parse", "head")
+	out, err := cmd.Output()
+	if err != nil {
+		return "", err
+	}
+	validTagRegexp, err := regexp.Compile("[^-_.a-zA-Z0-9]+")
+	if err != nil {
+		return "", err
+	}
+	sanitizedOutput := validTagRegexp.ReplaceAllString(string(out), "")
+	return sanitizedOutput, nil
+}
+
+// run prepares command execution and handles gathering logs to file
+func run(o options, name, repo, tag string, subs map[string]string) error {
+	var s []string
+
+	s = append(s,
+		parseVariable("_TAG", tag),
+		parseVariable("_REPOSITORY", repo),
+	)
+
+	// parse additional substitutions
+	for k, v := range subs {
+		s = append(s, parseVariable(k, v))
+	}
+	args := []string{
+		"builds", "submit",
+		"--config", o.configFile,
+		"--substitutions", strings.Join(s, ","),
+	}
+
+	if o.project != "" {
+		args = append(args, "--project", o.project)
+	}
+	// TODO (@Ressetkk): custom staging bucket implementation and re-using source code in builds
+	//if o.stagingBucket != "" {
+	//
+	//}
+
+	cmd := exec.Command("gcloud", args...)
+
+	var outw []io.Writer
+	var errw []io.Writer
+
+	if !o.silent {
+		outw = append(outw, os.Stdout)
+		errw = append(errw, os.Stderr)
+	}
+
+	f, err := os.Create(filepath.Join(o.logDir, strings.TrimSpace("build_"+strings.TrimSpace(name)+".log")))
+	if err != nil {
+		return fmt.Errorf("could not create log file: %w", err)
+	}
+
+	outw = append(outw, f)
+	errw = append(errw, f)
+
+	cmd.Stdout = io.MultiWriter(outw...)
+	cmd.Stderr = io.MultiWriter(errw...)
+
+	if err := cmd.Run(); err != nil {
```

Comment:
Why you call a gcloud cli command and not use API?

Reply:
Because of various reasons:
* We lose the ability to use cloudbuild.yaml - I'd have to write my own parser for custom cloudbuild-like file.
* We can ensure that we use updated gcloud-sdk with Dependabot - see Dockerfile
* It's way easier to implement ready-made tool and populate it with some changes